docs/development/maintaining-dockerfile.md

```diff
@@ -0,0 +1,46 @@
+# Maintaining Dockerfile
+
+## Base Image
+
+The Base Image should be an official Python distribution. We should prefer the most
+recent release, when possible.
+
+## OS version
+
+The OS should be debian, as debian/ubunto are widely used at KBase. The version should
+be the current stable. The LTS version, which is often several versions behind stable,
+does not seem to be supported by recent Python distributions. That is, to get the most
+recent Python versions requires we can't use LTS, and conversely if we wished to use the
+LTS debian we wouldn't be able to use the most recent Python releases.
+
+It is more important to have a current Python, than a current OS. The OS really isn't
+much of a concern to the service, as long as the required OS-level dependencies are
+available. Python itself takes care of the interface to the OS, and that is all we are
+concerned with.
+
+## OS dependencies
+
+The OS dependencies are indicated in the Dockerfile as exact versions. This ensures that
+images are consistent, even as the base image evolves over time.
+
+## Different Dockerfiles
+
+At present, there are three Dockerfiles which must be kept in synchrony. They differ
+enough that the inconvenience of keeping them consistent seems worth the effort.
+
+- `./Dockerfile`
+  - used for the production image
+  - needs to copy all service files
+  - does not include development Python dependencies
+  - has the production entrypoint
+- `./.devcontainer/Dockerfile`
```

Comment:
I think we could probably use one dockerfile and I think we could copy in the helper scripts and alternate endpoints in there as well. The entrypoint could also install required dev dependencies as well.

Reply:
Ugh, I was hoping not to have to deal with multipurposing a single image.
What about using a Dockerfile include for the main setup? Basically the base could be from the base image up to the installation of Python service dependencies. This basically leaves the main service image to set up the entrypoint, and the development images to install the dev dependencies and have separate default entrypoints?

Reply:
Well, I had read about Dockerfile INCLUDE, and not tried it ... I now have, and it isn't working (but perhaps that is just me, atm.) It also relies upon 3rd party extensions. The Docker team does not want to implement it, despite passionate and salty pleas: https://github.com/moby/moby/issues/735

Reply:
Alternative way is
```
# Base stage
FROM python:3.11 AS base
# Install shared dependencies, set up common environment...

# Production stage
FROM base AS production
# Copy service files, set up production-specific stuff...

# Development stage
FROM base AS development
# Set up dev-specific stuff...
```

docker build --target base -t myimage:base .
docker build --target production -t myimage:production .
docker build --target development -t myimage:development .


Could also use  --build-args as well

Reply:
Nice! 
I'll have a crack at this, though I'd do it using docker compose files.

Reply:
Okay, I've got this working locally.
Since I'm touching the Dockerfile with an eye to simplicity , I've been puzzled by some aspects of the codebase that aren't explained in any documents or comments.

A. I'm wondering why the entire repo is copied into /kb/module, rather than simply the components that are needed to run the service.
B. And I'm wondering why some bits ARE copied to /kb/deployment, when they are already present in /kb/module and could just be used from there (okay, I understand that one wants a more controlled deployment environment, but then why A)
C. And I'm wondering how the bits in the codebase are used that are not used in the service. E.g. scripts like scripts/prune_acls.py, refresh_token.py, or the entire top level import_specifications directory.
D. And there is deployment stuff like the top level hooks and build directories which seems defunct with GHA; unused? used in some other context?
E. Why is there an auth url in /deployment/conf/deployment.cfg that points ci? Presumably in prod the deployment config is volume mounted and KB_DEPLOYMENT_CONFIG is set to that path, and the one in the codebase is used just for local containers and possibly for a CI deployment.

Reply:
Oh, and another question -
F. Wondering why the image installs htop and wget -- as conveniences for poking around a container? If they aren't used I'd remove them and only install in the dev target.

Reply:
okay, @bio-boris , the Dockerfile and related changes are up now. I also took the opportunity to fold in some tool (black, isort, autoflake, flake8, mypy) improvements.

Reply:
```
3.200 Package wget is not available, but is referred to by another package.
3.200 This may mean that the package is missing, has been obsoleted, or
3.200 is only available from another source
3.200
3.202 E: Version '1.21.3-1+b2' for 'wget' was not found
------
WARNING: No output specified with docker-container driver. Build result will only remain in the build cache. To push result image into registry use --push or to load image into docker use --load
Dockerfile:26
--------------------
  25 |     # Install OS dependencies required by or nice-to-have in a development image
  26 | >>> RUN apt-get update && \
  27 | >>>     apt-get install -y --no-install-recommends htop=3.2.2-2 wget=1.21.3-1+b2 git=1:2.39.2-1.1 openssh-client=1:9.2p1-2 && \
  28 | >>>     apt-get clean && \
  29 | >>>     rm -rf /var/lib/apt/lists/*
  30 |     # Install Python dependencies require by development tools (cli tools and devcontainer)
--------------------
ERROR: failed to solve: process "/bin/sh -c apt-get update &&     apt-get install -y --no-install-recommends htop=3.2.2-2 wget=1.21.3-1+b2 git=1:2.39.2-1.1 openssh-client=1:9.2p1-2 &&     apt-get clean &&     rm -rf /var/lib/apt/lists/*" did not complete successfully: exit code: 100
docker build . --target dev
```

Reply:
Its very convenient to have wget and curl in all containers. They are very small and are very useful and also useful, especially when we get network connectivity issues.  I'd prefer not to remove that from the main image, as we can always install htop. In the distant, but not too distant future, we can use k8 debug containers, but until then it saves a lot of trouble by having them directly in the container. In fact, we could even use DockerSlim once we can use debug containers.

Reply:
Docs about some of those file are in our confluence

```
****************************
ACL And File Manager CronJob 
****************************

/root/bulk/acl_manager/ contains the required globus.cfg and python dependencies to run *prune_and_delete.py* 

The threshold to delete files and acls is defined as the constant THRESHOLD_DAYS, (currently set to 60)

If the modfied time of the user's directory is > THRESHOLD_DAYS, the ACL is marked as old, and deleted from globus, as well as the entire user's directory on */dtn/disk0/bulk/<username>*

The script should be run in the /root/bulk/acl_manager directory in order to access the *globus.cfg* and to consistently save logs in the same place.


See `here. <prune_and_delete.py>`_ 


****************************
ISSUES
****************************
* run /opt/kb-ftp-api/refresh_token.py to refresh token manually, otherwise this is run in a cronojb
* if it fails, uncomment the code inside to request a new token and rerun it
* if the staging service token fails, copy over the "REFRESH TOKENS" only into the appropriate values into the staging services globus.cfg file.
*

```